src/Catalog/Monitoring/Validation/Test/ValidationTestUtility.cs

```diff
@@ -0,0 +1,225 @@
+        private class BoolComparer : IEqualityComparer<object>
+        {
+            public new bool Equals(object x, object y)
+            {
+                var boolX = (bool)x;
+                var boolY = (bool)y;
+
+                return boolX == boolY;
+            }
+
+            public int GetHashCode(object obj)
+            {
+                return obj.GetHashCode();
+            }
+        }
+
+        /// <summary>
+        /// Returns true
+        ///   if both <see cref="string"/>s are null or empty
+        ///   or if <see cref="string.Equals(string)"/> returns true.
+        ///   
+        /// Returns false otherwise.
+        /// </summary>
+        private class StringComparer : IEqualityComparer<object>
+        {
+            public new bool Equals(object x, object y)
+            {
+                var stringX = (string)x;
+                var stringY = (string)y;
+
+                if (string.IsNullOrEmpty(stringX))
+                {
+                    return string.IsNullOrEmpty(stringY);
+                }
+
+                return string.Equals(stringX, stringY);
+            }
+
+            public int GetHashCode(object obj)
+            {
+                return obj.GetHashCode();
+            }
+        }
+
+        /// <summary>
+        /// Returns true
+        ///   if both <see cref="DateTimeOffset"/>s are null or <see cref="DateTimeOffset.MinValue"/>
+        ///   or if both <see cref="DateTimeOffset.Equals(DateTimeOffset)"/> returns true.
+        ///   
+        /// Returns false otherwise.
+        /// </summary>
+        private class DateTimeOffsetComparer : IEqualityComparer<object>
+        {
+            private bool DateTimeOffsetIsNullOrMin(DateTimeOffset date)
+            {
+                return date == null || date == DateTimeOffset.MinValue;
+            }
+            
+            public new bool Equals(object x, object y)
+            {
+                var dateX = (DateTimeOffset) x;
+                var dateY = (DateTimeOffset) y;
+
+                dateX = dateX.ToUniversalTime();
+                dateY = dateY.ToUniversalTime();
+
+                if (x == null || y == null)
+                {
+                    return DateTimeOffsetIsNullOrMin(dateX) && DateTimeOffsetIsNullOrMin(dateY);
+                }
+
+                return DateTimeOffset.Equals(dateX, dateY);
+            }
+
+            public int GetHashCode(object obj)
+            {
+                return obj.GetHashCode();
+            }
+        }
+
+        /// <summary>
+        /// Casts <paramref name="x"/> and <paramref name="y"/> to a <see cref="T"/> and returns <see cref="T.Equals(object)"/>.
+        /// </summary>
+        private class CastingComparer<T> : IEqualityComparer<object> where T : class
+        {
+            public new bool Equals(object x, object y)
+            {
+                var objX = (T)x;
+                var objY = (T)y;
+
+                return objX == objY;
```

Comment:
I think I used `.Equals` initially and ran into issues, but I'll look into it again.